src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+    }
+
+    // TODO: Use G_DBUS_ERROR_UNKNOWN_PROPERTY once we get glib >= 2.41.0
+    g_set_error(error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "Cannot get property '%s'", property_name);
+    return NULL;
+}
+
+gboolean upgrade_factory_snapshot_set_property_cb(GDBusConnection *connection, const gchar *sender,
+        const gchar *object_path, const gchar *interface_name, const gchar *property_name,
+        GVariant *value, GError **error, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    // TODO: Use G_DBUS_ERROR_PROPERTY_READ_ONLY and G_DBUS_ERROR_UNKNOWN_PROPERTY once we
+    // upgrade to glib >= 2.41.0 that has these new defines (added in glib commit 76d6fd0)
+    g_set_error(error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "Cannot set property '%s'", property_name);
+
+    return FALSE;
+}
+
+void upgrade_factory_snapshot_bus_acquired_cb(GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    ufs->system_bus = connection;
+    g_object_ref(ufs->system_bus);
+
+    GError *error = NULL;
+    GDBusNodeInfo *node_info = g_dbus_node_info_new_for_xml(""
+        "<!DOCTYPE node PUBLIC \"-//freedesktop//DTD D-BUS Object Introspection 1.0//EN\"\n"
+        "  \"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd\">\n"
+        "<node name=\"" UFS_DBUS_PATH "\">\n"
+        "  <interface name=\"" UFS_DBUS_INTERFACE "\">\n"
+        "    <method name=\"Start\">\n"
+        "      <arg name=\"release\" type=\"s\" direction=\"in\" />\n"
+        "      <arg name=\"result\" type=\"b\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <method name=\"GetProgress\">\n"
+        "      <arg name=\"queue_name\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"task_name\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"queue_pos\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"queue_total\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"partition_name\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"partition_pos\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"partition_total\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"message\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"progress\" type=\"i\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <signal name=\"ProgressChanged\">\n"
+        "    </signal>\n"
+        "    <property name=\"running\" type=\"b\" access=\"read\">\n"
+        "    </property>\n"
+        "  </interface>\n"
+        "</node>\n"
+    "", &error);
+
+    if (!node_info) {
+        SFMF_FAIL("Could not compile D-Bus XML: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    ufs->object_registration = g_dbus_connection_register_object(ufs->system_bus, UFS_DBUS_PATH,
+            g_dbus_node_info_lookup_interface(node_info, UFS_DBUS_INTERFACE),
```

Comment:
Yes, but then the next reader will wonder if there is a memory leak... which can be avoided by adding a comment... at which point it might as well be macro that returns + explicit exit...

Or worse, when doing changes later it gets mistaken for "just error reporting" and then we have a daemon that exits on some possibly rare but non fatal glitch.

At some point in time I came to the conclusion that macros that hide exits/returns are evil unless they have a name that makes it very clear what is going to happen, like: xxx_and_exit, xxx_die, xxx_abort.